### DISCUSSION TITLE
Which DpiAwareness is effective for UDF?

### DISCUSSION THREAD

**Discussion by fdzatdtv** [MicrosoftEdge/WebView2Feedback#4136] [Q&A] [answered]
I know the UDF (user-data-folder) is dependent on the DPI awareness and only works for a single DPI awareness. If you need to work with different, you need a separate UDF for each one.

I have lot of exceptions with HRESULT 0x8007139F although I'm using different UDFs. But how is Microsoft determining the effective DPI awareness used in conjunction with the UDF?
Is it **GetProcessDpiAwareness**? Or is it **GetThreadDpiAwarenessContext**?

**Comment by victorhuangwq** [comment]
@johna-ms could you help @fdzatdtv out with his question?

**Reply by victorhuangwq** [reply]
actually, it seems like @LiangTheDev might have a better understanding of this.

**Reply by LiangTheDev** [reply, marked as answer]
GetWindowDpiAwarenessContext, and fallback to GetProcessDpiAwareness if GetWindowDpiAwarenessContext is not supported for that version of the OS.
Note that there are app compat settings that could forcefully alter the dpiawareness of the processes.

**Reply by victorhuangwq** [reply]
@fdzatdtv ^

**Reply by fdzatdtv** [reply]
@LiangTheDev When is GetWindowDpiAwarenessContext called? Is it during EnsureCoreWebView2Async?

**Reply by LiangTheDev** [reply]
Yes, during EnsureCoreWebView2Async.

**Comment by fdzatdtv** [comment]
Any news on that?

What if the DpiAwareness is Unaware at the beginning but changes later to Aware?

**Reply by fdzatdtv** [reply]
We have a bunch of processes using WebView2. And want them to use as less UDF as possible.  We make the UDF dependent of the Dpi awareness and the culture. Yet, we are calling GetProcessDpiAwareness to determine the awareness.

But in many cases our processes run into a HRESULT 0x8007139F. We've proved that this is not due to fast initialization of several instances nor fast dispose of last and init of new one.

So, we highly suspect that it has to do with the Dpi Awareness. Probably, the dpi awareness changes during runtime, after we have determined with GetProcessDpiAwareness, but before you stick the UDF to the dpi awareness you determined.

Having the UDF path, is there any way to (programmatically) determine to which dpi awareness it is bound to? @LiangTheDev

**Reply by LiangTheDev** [reply]
There is currently no API for doing that. The dpi awareness is represented on WebView2 browser process's command line as --embedded-browser-webview-dpi-awareness switch. You could potentially query the command line of the browser process. The browser process id can be obtained from BrowserProcessId property.

If you attach debugger to the app process, it should print out some messages about why the creation failed. What have you observed?

**Reply by fdzatdtv** [reply]
The exceptions happen (only) on production systems. I can't attach a debugger there. But thank you for the advice. I'll query the command line of all msedgewebview2 processes with the same UDF at the time of the exception and write to a log file.

The log file is sent automatically to us. Thus, I'm able to analyse further.

**Reply by fdzatdtv** [reply]
@LiangTheDev Can you tell me what the integer values of the --embedded-browser-webview-dpi-awareness switch correspond to?

**Comment by LiangTheDev** [comment]
It should be the values of DPI_AWARENESS https://learn.microsoft.com/en-us/windows/win32/api/windef/ne-windef-dpi_awareness
